appbuilder/services/npm-service.ts

```diff
@@ -0,0 +1,250 @@
+import * as path from "path";
+import * as os from "os";
+import * as constants from "../../constants";
+import { fromWindowsRelativePathToUnix } from "../../helpers";
+import { exportedPromise } from "../../decorators";
+
+export class NpmService implements INpmService {
+	private static NPM_MODULE_NAME = "npm";
+	private static TYPES_DIRECTORY = "@types/";
+	private static TNS_CORE_MODULES_DEFINITION_FILE_NAME = `${constants.TNS_CORE_MODULES}${constants.FileExtensions.TYPESCRIPT_DEFINITION_FILE}`;
+	private static NPM_REGISTRY_URL = "https://registry.npmjs.org";
+
+	private _npmExecutableName: string;
+	private _npmBinary: string;
+
+	constructor(private $childProcess: IChildProcess,
+		private $errors: IErrors,
+		private $fs: IFileSystem,
+		private $hostInfo: IHostInfo,
+		private $httpClient: Server.IHttpClient,
+		private $logger: ILogger,
+		private $projectConstants: Project.IConstants) { }
+
+	private get npmBinary(): string {
+		if (!this._npmBinary) {
+			try {
+				require(NpmService.NPM_MODULE_NAME);
+				let npmMainJsFile = require.resolve(NpmService.NPM_MODULE_NAME);
+				let nodeModulesDirName = "node_modules";
+				this._npmBinary = path.join(npmMainJsFile.substring(0, npmMainJsFile.lastIndexOf(nodeModulesDirName) + nodeModulesDirName.length), ".bin", this.npmExecutableName);
```

Comment:
It may be sensible to check if the binary file exists or not. There may be a case where we don't have a `.bin` directory. I am unsure how to go about it then, though. It seems reasonable to throw an error if neither the `.bin`, nor a globally installed `npm` is present.
@TsvetanMilanov any thoughts on this?